controller/pkg/finalizer/handlerror.go

```diff
@@ -0,0 +1,49 @@
+package finalizer
+
+import (
+	"github.com/go-logr/logr"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+
+	ctrlconsts "github.com/kong/kong-operator/controller/consts"
+	"github.com/kong/kong-operator/controller/pkg/log"
+)
+
+// HandlePatchOrUpdateError handles errors returned from patch or update operations
+// on Kubernetes resources when changing finalizers.
+func HandlePatchOrUpdateError(err error, logger logr.Logger) (ctrl.Result, error) {
```

Comment:
I think we can add a `resourceType` string parameter to provide better context. e.g. `Gateway`, `Controlplane` etc

WDYT?

Reply:
The original log is in a format similar to this, which includes the resource type.
```
if k8serrors.IsConflict(err) {
					log.Debug(logger, "conflict found when updating ControlPlane, retrying")
					return ctrl.Result{Requeue: true, RequeueAfter: ctrlconsts.RequeueWithoutBackoff}, nil
				}
				return ctrl.Result{}, fmt.Errorf("failed updating ControlPlane: %w", err)
```